[cypress] add SQL lab tests

williaster opened the pull request:
This PR adds new cypress tests for SQL lab

**SQL tabs**
1) supports creating new tab
2) supports removing a tab

**Table selector**
1) renders dropdowns to select database, schema, and table
2) asserts that you can sequentially select `database > schema > table`, and that doing so 
  a) generates a table schema preview 
  b) returns a dataset preview

**Query panel**
1) supports entering and running a query
2) supports saving a query and visiting the saved query url to assert the same query + results 
   (also tests `ctrl+r` keyboard shortcut for running a query)

Other tests I considered but couldn't implement easily due to Cypress limitations, etc.
- renaming a tab (requires browser dialog support)
- downloading csv 

@kristw @michellethomas @graceguo-supercat @mistercrunch

williaster committed:
[cypress] add sqllab/sourcePanel integration tests

williaster committed:
[cypress] add tests for SQL Lab query panel

williaster committed:
[cypress] add selectors/

williaster committed:
[cypress] add tests for SQL Lab tabs

kristw commented on superset/assets/cypress/integration/sqllab/query.js:
```diff
@@ -0,0 +1,125 @@
+      )
+      .then(() => {
+        cy.get('#js-sql-toolbar button')
+          .eq(0)
```
what is `eq(0)`

williaster replied:
takes the first element, generalized version of `first()` or `last()`

kristw commented on superset/assets/cypress/integration/sqllab/query.js:
```diff
@@ -0,0 +1,125 @@
+                .click()
+                .then(() => {
+                  // Enter name + save into modal
+                  cy.get('.modal-sm input')
```
some of these nested `then` clause may not needed to be nested. 
Cypress does some magic to apply `.then` for you under the hood.

```js
cy.get('a').click()
cy.get('b').click() // will wait for the statement above to resolve
```

https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Promises

williaster replied:
will play with this 👍 cypress is magical ✨

williaster replied:
@michellethomas @kristw did you have any conclusions whether nested `then()`s can mitigate flakiness?

kristw commented on superset/assets/cypress/selectors/index.js:
```diff
@@ -0,0 +1 @@
+export const selectResultsTab = () => cy.get('.SouthPane .ReactVirtualized__Table');
```
Why is this one in a separate file?

williaster replied:
I thought it could possibly be useful to add a concept of re-usable `selectors` that are used in multiple tests and select specific elements. The alternatives would be 

- add a cypress command to select that element (seemed too heavy for this one in particular)
- write the same selector multiple times (this could be fine, if you need to change them they'd be changed in multiple places)

(`"selector"` was borrowed the word from  a common pattern in redux apps to select specific data from the state tree). totally happy to remove this if you + @michellethomas aren't into it.

kristw replied:
I agree with not writing selector string multiple times. More thinking about where to put the function. My main concern is avoiding a setup that leads to one exploding file that contains selectors from every part of Superset. From the name of the file now it invites people to add selector from all other pages here. Other options would be 
- Rename this to `sqlLab.js` to scope it a bit.
- Put the function in test file itself.

I am more comfortable with either option above.

williaster committed:
[cypress][sqllab] remove nested assertions where possible, update to single test index file

williaster committed:
[cypress][sqllab] try force click

williaster committed:
[cypress][sqllab] wait for saved query to load

williaster committed:
[cypress][sqllab] unnest, use visit for href

williaster committed:
[cypress][sqllab] try 1s wait :O

williaster commented on superset/assets/cypress/integration/sqllab/_query.js:
```diff
@@ -0,0 +1,97 @@
+        .should('have.attr', 'href')
+        .then((href) => {
+          cy.visit(href).then(() => {
+            cy.wait(1000);
```
this was the only way I could get this test to pass 😭 the saved query page seems to take some time to load, and without this I get errors for the run query button having a size of `0 x 0`

@michellethomas do you have any thoughts on other approaches that wouldn't require a `wait`?

michellethomas replied:
You can also try to use the timeout option on get if you want to extend the amount of time it takes before it times out.

williaster committed:
[cypress][sqllab] wait for savedquery api response

williaster committed:
[cypress][sqllab] fix wait syntax

williaster committed:
[cypress][sqllab] add route wait in sourcepanel for flakiness

williaster committed:
[cypress] use *.test.js pattern for tests

williaster committed:
[cypress] fix paths

williaster committed:
[cypress][sqllab] actually try running lint

williaster committed:
[cypress][dashboards] use *.test.js pattern for tests

williaster commented:
@kristw @michellethomas @graceguo-supercat I think this is good to go now 👍less flaky, removed nested tests where it seemed to make sense, and I updated all the test patterns for visualizations + dashboards. can you PTAL? 👀

kristw commented on superset/assets/cypress/integration/sqllab/tabs.js:
```diff
@@ -0,0 +1,39 @@
+        // open the tab dropdown to remove
+        cy.get('#a11y-query-editor-tabs > ul > li:first button')
+          .click()
+          .then(() => {
```
Is this `.then` required?

kristw commented on superset/assets/cypress/integration/sqllab/sourcePanel.js:
```diff
@@ -0,0 +1,60 @@
+      cy.visit('/superset/sqllab');
+    });
+
+    it('renders dropdowns to select database, schema, and table', () => {
```
for a small test like this, perhaps merge with the following test to save runtime. 
Can also use `should('have.length')` instead of `.then( expect )`.

kristw reviewed (approved):
A few minor comments but overall LGTM.

williaster committed:
[cypress][sqllab] combine tests, remove nesting

williaster committed:
[cypress][sqllab] explicitly wait for query results
